## Homework 3

> [!NOTE]
> This homework uses the pinned 2026 lead-scoring release in the course
> repository. The plan and report are available in `cohorts/2026/data/`.

### Dataset

In this homework, we will use the 2026 lead-scoring dataset. Download it from [here](https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/course_lead_scoring_2026.csv).

Or you can do it with `wget`:

```bash
wget https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/course_lead_scoring_2026.csv
```

In this dataset our desired target for classification task will be `converted` variable - has the client signed up to the platform or not.

### Data preparation

* Check if the missing values are presented in the features.
* If there are missing values:
    * For categorical features, replace them with 'NA'
    * For numerical features, replace with with 0.0 


In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [24]:
df = pd.read_csv('/Users/foroghakbari/Zoomcamp/ml_course/zoomcamp_ml/classification/course_lead_score.csv')

In [25]:
df.head().T

,0,1,2,3,4
lead_source,organic_search,social_media,referral,paid_ads,referral
industry,technology,technology,retail,manufacturing,manufacturing
employment_status,employed,employed,employed,student,student
location,europe,south_america,europe,NaN,north_america
annual_income,88160.0,72688.0,44697.0,NaN,24062.0
number_of_courses_viewed,3,3,3,3,4
interaction_count,4,7,4,6,5
lead_score,0.64,0.72,0.58,0.57,0.62
converted,1,1,1,0,1


In [26]:
df.columns = df.columns.str.lower().str.replace(' ', '_')

In [27]:
#fill NAs
# Create a fill values dictionary
fill_dict = {
    col: 0.0 if pd.api.types.is_numeric_dtype(dtype) else 'NA'
    for col, dtype in df.dtypes.items()
}

# Apply fillna in one step
df = df.fillna(value=fill_dict)


In [28]:
# Total missing values per column
print(df.isna().sum())

lead_source                 0
industry                    0
employment_status           0
location                    0
annual_income               0
number_of_courses_viewed    0
interaction_count           0
lead_score                  0
converted                   0
dtype: int64


In [30]:
df

,lead_source,industry,employment_status,location,annual_income,number_of_courses_viewed,interaction_count,lead_score,converted
0,organic_search,technology,employed,europe,88160.0,3,4,0.64,1
1,social_media,technology,employed,south_america,72688.0,3,7,0.72,1
2,referral,retail,employed,europe,44697.0,3,4,0.58,1
3,paid_ads,manufacturing,student,NA,0.0,3,6,0.57,0
4,referral,manufacturing,student,north_america,24062.0,4,5,0.62,1
...,...,...,...,...,...,...,...,...,...
4995,organic_search,education,employed,asia,45834.0,2,2,0.37,1
4996,NA,technology,employed,north_america,84831.0,2,5,0.51,0
4997,organic_search,finance,employed,asia,53636.0,4,6,0.75,1
4998,referral,technology,employed,africa,69430.0,4,8,0.79,1


In [29]:
df.dtypes

lead_source                  object
industry                     object
employment_status            object
location                     object
annual_income               float64
number_of_courses_viewed      int64
interaction_count             int64
lead_score                  float64
converted                     int64
dtype: object

### Question 1

What is the most frequent observation (mode) for the column `industry`?

- `NA`
- `technology`
- `healthcare`
- `retail`

In [7]:
df['industry']

0          technology
1          technology
2              retail
3       manufacturing
4       manufacturing
            ...      
4995        education
4996       technology
4997          finance
4998       technology
4999       technology
Name: industry, Length: 5000, dtype: object

In [9]:
df.industry.value_counts()

industry
technology       1173
retail            925
healthcare        840
finance           720
education         639
manufacturing     463
Name: count, dtype: int64

### Question 2

Create the [correlation matrix](https://www.google.com/search?q=correlation+matrix) for the numerical features of your dataset.
In a correlation matrix, you compute the correlation coefficient between every pair of features.

What are the two features that have the biggest correlation?

- `interaction_count` and `lead_score`
- `number_of_courses_viewed` and `lead_score`
- `number_of_courses_viewed` and `interaction_count`
- `annual_income` and `interaction_count`

Only consider the pairs above when answering this question.

### Split the data

- Split the data with these exact calls:

```python
df_full_train, df_test = train_test_split(df, test_size=0.2, random_state=42)
df_train, df_val = train_test_split(
    df_full_train, test_size=0.25, random_state=42
)
```

- Make sure that the target value `converted` is not in your dataframe.


In [10]:
df.dtypes

lead_source                  object
industry                     object
employment_status            object
location                     object
annual_income               float64
number_of_courses_viewed      int64
interaction_count             int64
lead_score                  float64
converted                     int64
dtype: object

In [11]:
# Alternatively, specify exact types:
numeric_df = df.select_dtypes(include=['int', 'float'])

In [12]:
numeric_df

,annual_income,number_of_courses_viewed,interaction_count,lead_score,converted
0,88160.0,3,4,0.64,1
1,72688.0,3,7,0.72,1
2,44697.0,3,4,0.58,1
3,NaN,3,6,0.57,0
4,24062.0,4,5,0.62,1
...,...,...,...,...,...
4995,45834.0,2,2,0.37,1
4996,84831.0,2,5,0.51,0
4997,53636.0,4,6,0.75,1
4998,69430.0,4,8,0.79,1


In [15]:
import numpy as np
import pandas as pd

# Your existing code
corr_matrix = df.corr(numeric_only=True)

# Mask the lower triangle and diagonal (self-correlations)
upper_triangle = corr_matrix.where(
    np.triu(np.ones(corr_matrix.shape), k=1).astype(bool)
)

# Stack into a Series, sort by absolute value, and grab top results
top_correlations = (
    upper_triangle.stack()
    .reindex(upper_triangle.stack().abs().sort_values(ascending=False).index)
)

print("Top Correlations:")
print(top_correlations.head(5))

Top Correlations:
interaction_count         lead_score           0.928712
number_of_courses_viewed  lead_score           0.769352
                          interaction_count    0.721609
lead_score                converted            0.490206
interaction_count         converted            0.448624
dtype: float64


In [17]:
df

,lead_source,industry,employment_status,location,annual_income,number_of_courses_viewed,interaction_count,lead_score,converted
0,organic_search,technology,employed,europe,88160.0,3,4,0.64,1
1,social_media,technology,employed,south_america,72688.0,3,7,0.72,1
2,referral,retail,employed,europe,44697.0,3,4,0.58,1
3,paid_ads,manufacturing,student,NaN,NaN,3,6,0.57,0
4,referral,manufacturing,student,north_america,24062.0,4,5,0.62,1
...,...,...,...,...,...,...,...,...,...
4995,organic_search,education,employed,asia,45834.0,2,2,0.37,1
4996,NaN,technology,employed,north_america,84831.0,2,5,0.51,0
4997,organic_search,finance,employed,asia,53636.0,4,6,0.75,1
4998,referral,technology,employed,africa,69430.0,4,8,0.79,1


In [31]:
from sklearn.model_selection import train_test_split

In [32]:
df_full_train, df_test = train_test_split(df, test_size=0.2, random_state=42)
df_train, df_val = train_test_split(
    df_full_train, test_size=0.25, random_state=42
)

Make sure that the target value converted is not in your dataframe.

In [33]:
df_full_train

,lead_source,industry,employment_status,location,annual_income,number_of_courses_viewed,interaction_count,lead_score,converted
4227,events,healthcare,employed,south_america,47301.0,5,7,0.69,0
4676,organic_search,technology,unemployed,asia,0.0,1,5,0.57,0
800,events,finance,unemployed,asia,47970.0,0,0,0.17,0
3671,organic_search,retail,self_employed,south_america,52847.0,3,3,0.45,0
4193,paid_ads,finance,employed,africa,73460.0,1,4,0.42,1
...,...,...,...,...,...,...,...,...,...
4426,paid_ads,healthcare,self_employed,europe,49442.0,0,3,0.32,1
466,social_media,finance,student,north_america,0.0,1,1,0.23,0
3092,referral,healthcare,student,asia,16378.0,2,5,0.55,1
3772,social_media,NA,NA,north_america,87681.0,2,6,0.76,1


In [34]:
del df_train['converted']

In [35]:
del df_val['converted']
del df_test['converted']


In [36]:
df_train

,lead_source,industry,employment_status,location,annual_income,number_of_courses_viewed,interaction_count,lead_score
4702,paid_ads,finance,employed,asia,65980.0,2,3,0.37
2155,events,finance,employed,south_america,71662.0,0,0,0.09
289,referral,NA,employed,asia,40220.0,2,5,0.55
1822,organic_search,finance,employed,north_america,88352.0,2,2,0.45
3442,events,technology,employed,north_america,89316.0,1,5,0.00
...,...,...,...,...,...,...,...,...
3200,paid_ads,finance,employed,asia,81522.0,1,3,0.29
617,social_media,retail,employed,asia,33075.0,1,1,0.30
1992,organic_search,finance,employed,africa,71532.0,3,5,0.58
3301,social_media,retail,employed,asia,48836.0,0,0,0.03


### Question 3 - EDA

- Calculate the mutual information score between `converted` and other categorical variables in the dataset. Use the training set only.
- Round the scores to 2 decimals using `round(score, 2)`.

Which of these variables has the biggest mutual information score?

- `industry`
- `location`
- `lead_source`
- `employment_status`


### Question 4

- Now let's train a logistic regression.
- Remember that we have several categorical variables in the dataset. Include them using one-hot encoding.
- Fit the model on the training dataset.
  - To make sure the results are reproducible across different versions of Scikit-Learn, fit the model with these parameters:
  - `model = LogisticRegression(solver='liblinear', C=1.0, max_iter=1000, random_state=42)`
- Calculate the accuracy on the validation dataset and round it to 2 decimal digits.

What accuracy did you get?

- 0.55
- 0.65
- 0.75
- 0.85




### Question 5

- Let's find the least useful feature using the _feature elimination_ technique.
- Train a model using the same features and parameters as in Q4 (without rounding).
- Now exclude each feature from this set and train a model without it. Record the accuracy for each model.
- For each feature, calculate the difference between the original accuracy and the accuracy without the feature.

Which of following feature has the smallest difference?

- `'lead_source'`
- `'number_of_courses_viewed'`
- `'interaction_count'`

> **Note**: The difference doesn't have to be positive.



### Question 6

- Now let's train a regularized logistic regression.
- Let's try the following values of the parameter `C`: `[0.000001, 0.00001, 0.0001, 0.001]`.
- Train models using all the features as in Q4.
- Calculate the accuracy on the validation dataset and round it to 3 decimal digits.

Which of these `C` leads to the best accuracy on the validation set?

- 0.000001
- 0.00001
- 0.0001
- 0.001
